MapleServer2/Tools/MobAIManager.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.IO;
+using System.Collections.Generic;
+using System.Xml;
+using System.Xml.Schema;
+using MapleServer2.Types;
+
+namespace MapleServer2.Tools
+{
+    public static class MobAIManager
+    {
+        private static readonly Dictionary<string, MobAI> AiTable = new Dictionary<string, MobAI>();
+
+        public static void Load(string dirPath, string schemaPath = "")
+        {
+            foreach (string filePath in Directory.GetFiles(dirPath))
+            {
+                if (!filePath.EndsWith(".xml"))
+                {
+                    continue;
+                }
+
+                XmlDocument document = new XmlDocument();
+                document.Load(filePath);
+                if (schemaPath.Length > 0)
+                {
+                    document.Schemas.Add(null, schemaPath);
+                    try
+                    {
+                        document.Validate(null);
+                    }
+                    catch (XmlSchemaValidationException)
+                    {
+                        // TODO: log invalid AI files
+                        continue;
+                    }
+                }
+                ParseAI(document);
+                // TODO: log parsed AI files
+            }
+        }
+
+        public static MobAI GetAI(string aiInfo)
+        {
+            return AiTable.GetValueOrDefault(aiInfo, null);
+        }
+
+        private static void ParseAI(XmlDocument document)
+        {
+            XmlNode behaviorsNode = document.SelectSingleNode("/ai/behavior");
+
+            MobAI ai = new MobAI();
+            foreach (XmlNode node in behaviorsNode)
+            {
+                MobAction newActionValue = GetMobAction(node.Attributes["action"]?.Value);
+                MobMovement movementValue = GetMobMovement(node.Attributes["movement"]?.Value);
+
+                if (node.Name == "spawn")
+                {
+                    ai.Rules.Add(MobState.Spawn, (newActionValue, movementValue, Array.Empty<MobAI.Condition>()));
```

Comment:
Can we make the empty array the default and initialize it in the constructor?

Reply:
I'm not sure what you mean?

The empty array here though is temporary. My idea in the AI files was that would be able to specify <condition>'s during a certain state to limit an action.

This part would parse a list of Conditions (lambdas) and the AI would run the conditions when there's a chance to take an action.